File: src/main/scala/com/microsoft/ml/spark/cntk/train/CNTKLearner.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark.cntk.train

import java.io.File
import java.net.URI

import com.microsoft.ml.spark.cntk.CNTKModel
import com.microsoft.ml.spark.core.env.FileUtilities._
import com.microsoft.ml.spark.core.contracts.Wrappable
import com.microsoft.ml.spark.core.env.{EnvironmentUtils, FileUtilities, InternalWrapper}
import org.apache.hadoop.fs.Path
import org.apache.spark.ml._
import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
import org.apache.spark.ml.linalg.{DenseVector, SparseVector, Vector}
import org.apache.spark.ml.param._
import org.apache.spark.ml.util.{DefaultParamsReadable, Identifiable}
import org.apache.spark.sql._
import org.apache.spark.sql.types._

trait CNTKParams extends Params {

  val labelsColumnName = new Param[String](this, "labelsColumnName", "Label column name")
  def getLabelsColumnName: String = $(labelsColumnName)

  val featuresColumnName = new Param[String](this, "featuresColumnName", "Features column name")
  def getFeaturesColumnName: String = $(featuresColumnName)

  // This will go away after the CNTK HDFS Deserializer
  val localHdfsMount = new Param[String](this, "localHdfsMount", "Local mount point for hdfs:///")
  def getLocalHdfsMount: String = $(localHdfsMount)

  val dataTransfer = new Param[String](this, "dataTransfer", "Transfer strategy")
  def setTransferStrategy(s: String): this.type = set(dataTransfer, s)
  def getDataTransfer: String = $(dataTransfer)

  // TODO: Convert to enum contract shared with CNTK's HDFS Deserializer
  val dataFormat = new Param[String](this, "dataFormat", "Transfer format")
  def setDataFormat(s: String): this.type = set(dataFormat, s)
  def getDataFormat: String = $(dataFormat)

  val weightPrecision = new Param[String](this, "weightPrecision", "Weights")
  def getWeightPrecision: String = $(weightPrecision)

  val featureCount = new IntParam(this, "featureCount", "Number of features for reduction")
  def setFeatureCount(c: Int): this.type = set(featureCount, c)
  def getFeatureCount: Int = $(featureCount)

  val brainScript = new Param[String](this, "brainScript", "String of BrainScript config")
  def setBrainScriptText(t: String): this.type = set(brainScript, t)
  def setBrainScriptFile(f: String): this.type = set(brainScript, FileUtilities.readFile(new File(f)))
  def getBrainScript: String = $(brainScript)

  val parallelTrain = new BooleanParam(this, "parallelTrain", "Train using an MPI ring")
  def setParallelTrain(b: Boolean): this.type = set(parallelTrain, b)
  def getParallelTrain: Boolean = $(parallelTrain)

  val workingDir = new Param[String](this, "workingDir", "Working directory for CNTK")
  def setWorkingDirectory(d: String): this.type = set(workingDir, d)
  def getWorkingDirectory: String = $(workingDir)

  val gpuMachines = new StringArrayParam(this, "gpuMachines", "GPU machines to train on")
  setDefault(gpuMachines -> Array(s"127.0.0.1,${EnvironmentUtils.GPUCount.getOrElse(1)}"))
  def setGPUMachines(value: Array[String]): this.type = set(gpuMachines, value)
  def getGPUMachines: Array[String] = $(gpuMachines)

  val username = new Param[String](this, "username", "Username for the GPU VM")
  def setUserName(value: String): this.type = set(username, value)
  def getUserName: String = $(username)

}

object CNTKLearner extends DefaultParamsReadable[CNTKLearner] {
  val localDataTransfer = "local"
  val hdfsDataTransfer  = "hdfs"
  val textDataFormat    = "text"
  val parquetDataFormat = "parquet"
  val denseForm         = "dense"
  val sparseForm        = "sparse"
  val doublePrecision   = "double"
  val floatPrecision    = "float"
  val rpcPortNumber     = 8020
  val identityLocation  = "wasb:///MML-GPU/identity"
  val localSSH          = ".ssh/MML-GPU"
}

@InternalWrapper
class CNTKLearner(override val uid: String) extends Estimator[CNTKModel]
  with CNTKParams with Wrappable {

  setDefault(
    labelsColumnName-> "labels",
    featuresColumnName->"features",
    dataTransfer->"local",
    dataFormat->"text",
    weightPrecision-> CNTKLearner.floatPrecision,
    featureCount->1,
    parallelTrain->true,
    workingDir->"tmp",
    username->"sshuser"
  )

  def this() = this(Identifiable.randomUID("CNTKLearner"))

  override def fit(dataset: Dataset[_]): CNTKModel = {
    val spark = dataset.sparkSession
    val labels = getLabelsColumnName
    val features = getFeaturesColumnName

    // Convert label column to categorical on train, remove rows with missing labels
    val convertedLabelDataset = dataset.na.drop(Seq(labels))

    // The reduction step of Featurize
    val reducedData = DataTransferUtils.reduceAndAssemble(
      convertedLabelDataset,
      labels,
      features,
      getFeatureCount)

    // TODO: we should store vector sizes in schema for quick retrieval in the future
    val feature1 = reducedData.select(features).head.getAs[Vector](0)
    val featureDim = feature1.size
    val featureForm = feature1 match {
      case dv: DenseVector => CNTKLearner.denseForm
      case sv: SparseVector => CNTKLearner.sparseForm
    }

    val label1 = reducedData.select(labels).head.getAs[Vector](0)
    val labelDim = label1.size
    val labelForm = label1 match {
      case dv: DenseVector => CNTKLearner.denseForm
      case sv: SparseVector => CNTKLearner.sparseForm
    }

    val cntkrootURI = new URI(getWorkingDirectory)
    val cntkrootPath = new File(cntkrootURI).getAbsolutePath
    val relativeInPath = new Path(cntkrootPath, s"$uid-inputdata").toString
    logInfo(s"$uid working in $cntkrootPath with relative path $relativeInPath")

    val (writer, hdfsPath) = getWriter(dataset, relativeInPath)

    // Generating the CNTK text file
    val conformedData = getDataFormat match {
      case CNTKLearner.textDataFormat =>
        DataTransferUtils.convertDatasetToCNTKTextFormat(
          reducedData, labels, features, labelForm, featureForm)
      case CNTKLearner.parquetDataFormat =>
        DataTransferUtils.convertDatasetToCNTKParquetFormat(
          reducedData, labels, features, labelForm, featureForm, getWeightPrecision)
    }

    conformedData.persist()

    val remappedInPath = getDataFormat match {
      case CNTKLearner.textDataFormat => writer.checkpointToText(conformedData)
      case CNTKLearner.parquetDataFormat => writer.checkpointToParquet(conformedData)
    }
    val outputDir = s"$uid-outdir"
    val relativeOutRoot = s"$cntkrootPath/$outputDir"

    val config = new BrainScriptBuilder()
      .setOutputRoot(relativeOutRoot)
      // TODO: Refactor to more structured form of converting schema to CNTK config
      .setInputFile(
        remappedInPath,
        getDataFormat,
        Map(features -> InputShape(featureDim, featureForm),
            labels -> InputShape(labelDim, labelForm)))
      .setHDFSPath(hdfsPath)
      .setWeightPrecision(getWeightPrecision)

    // Train the learner
    val cb =
      if (getParallelTrain) new MPICommandBuilder(log, getGPUMachines, hdfsPath, remappedInPath, getUserName)
      else new CNTKCommandBuilder(log)
    cb.setWorkingDir(cntkrootPath)
      .insertBaseConfig(getBrainScript)
      .appendOverrideConfig(config.toOverrideConfig)
      .setOutputDir(outputDir)
      .setSparkSession(spark)
      .setDataFormat(getDataFormat)
      .setModelOutputDir(config.getLocalModelOutputDir)
      .setModelName(config.getModelName)

    cb.runCommand()
    logInfo("CNTK training finished")
    conformedData.unpersist()
    // Note: This currently runs only on linux
    new CNTKModel(uid + "-model")
      .setModelLocation(config.getModelPath)
      .setInputCol(features)
      .setOutputCol(labels)
  }

  def getWriter(dataset: Dataset[_], relativeInPath: String): (SingleFileResolver,
    Option[(String, String, String)]) = {
    getDataTransfer match {
      case CNTKLearner.localDataTransfer => (new LocalWriter(log, relativeInPath), None)
      case CNTKLearner.hdfsDataTransfer => {
        val mntpt = if (isDefined(localHdfsMount)) {
          val x = getLocalHdfsMount
          logInfo(s"Using override hdfsm point: $x")
          x
        } else {
          val x = sys.env.getOrElse("HDFS_MOUNTPOINT", "tmp/mnt")
          logInfo(s"Using deduced hdfsm point: $x")
          x
        }
        logInfo(s"hdfs-mount mounted at $mntpt, writing ${dataset.rdd.getNumPartitions} distributed files")
        val hdfsWriter =
          new HdfsWriter(log,
                         mntpt,
                         dataset.rdd.getNumPartitions,
                         relativeInPath,
                         dataset.sparkSession.sparkContext)
        val hdfsPath = Some((hdfsWriter.getHdfsInputDataDir,
                             hdfsWriter.getRootDir,
                             hdfsWriter.getNameNode))
        (hdfsWriter, hdfsPath)
      }
      case _ =>
        throw new Exception(
          s"Only ${CNTKLearner.localDataTransfer} and" +
          s" ${CNTKLearner.hdfsDataTransfer} supported options for data transfer")
    }
  }

  override def copy(extra: ParamMap): Estimator[CNTKModel] = defaultCopy(extra)

  /** Taken from CNTKModel - adds the output column to the input
    * @param schema The input schema
    * @return The transformed schema with output column added
    */
  override def transformSchema(schema: StructType): StructType =
    schema.add(getLabelsColumnName, VectorType)

}

File: src/src/main/resources/mmlspark/_CNTKLearner.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class _CNTKLearner(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """
    ``CNTKLearner`` trains a model on a dataset on a GPU edge node.  The
    result is a ``CNTKModel``.

    Args:

        brainScript (str): String of BrainScript config
        dataFormat (str): Transfer format (default: text)
        dataTransfer (str): Transfer strategy (default: local)
        featureCount (int): Number of features for reduction (default: 1)
        featuresColumnName (str): Features column name (default: features)
        gpuMachines (list): GPU machines to train on (default: [Ljava.lang.String;@17c8b8f0)
        labelsColumnName (str): Label column name (default: labels)
        localHdfsMount (str): Local mount point for hdfs:///
        parallelTrain (bool): Train using an MPI ring (default: true)
        username (str): Username for the GPU VM (default: sshuser)
        weightPrecision (str): Weights (default: float)
        workingDir (str): Working directory for CNTK (default: tmp)
    """

    @keyword_only
    def __init__(self, brainScript=None, dataFormat="text", dataTransfer="local", featureCount=1, featuresColumnName="features", gpuMachines=["127.0.0.1,1"], labelsColumnName="labels", localHdfsMount=None, parallelTrain=True, username="sshuser", weightPrecision="float", workingDir="tmp"):
        super(_CNTKLearner, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.CNTKLearner")
        self.brainScript = Param(self, "brainScript", "brainScript: String of BrainScript config")
        self.dataFormat = Param(self, "dataFormat", "dataFormat: Transfer format (default: text)")
        self._setDefault(dataFormat="text")
        self.dataTransfer = Param(self, "dataTransfer", "dataTransfer: Transfer strategy (default: local)")
        self._setDefault(dataTransfer="local")
        self.featureCount = Param(self, "featureCount", "featureCount: Number of features for reduction (default: 1)")
        self._setDefault(featureCount=1)
        self.featuresColumnName = Param(self, "featuresColumnName", "featuresColumnName: Features column name (default: features)")
        self._setDefault(featuresColumnName="features")
        self.gpuMachines = Param(self, "gpuMachines", "gpuMachines: GPU machines to train on (default: [Ljava.lang.String;@17c8b8f0)")
        self._setDefault(gpuMachines=["127.0.0.1,1"])
        self.labelsColumnName = Param(self, "labelsColumnName", "labelsColumnName: Label column name (default: labels)")
        self._setDefault(labelsColumnName="labels")
        self.localHdfsMount = Param(self, "localHdfsMount", "localHdfsMount: Local mount point for hdfs:///")
        self.parallelTrain = Param(self, "parallelTrain", "parallelTrain: Train using an MPI ring (default: true)")
        self._setDefault(parallelTrain=True)
        self.username = Param(self, "username", "username: Username for the GPU VM (default: sshuser)")
        self._setDefault(username="sshuser")
        self.weightPrecision = Param(self, "weightPrecision", "weightPrecision: Weights (default: float)")
        self._setDefault(weightPrecision="float")
        self.workingDir = Param(self, "workingDir", "workingDir: Working directory for CNTK (default: tmp)")
        self._setDefault(workingDir="tmp")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, brainScript=None, dataFormat="text", dataTransfer="local", featureCount=1, featuresColumnName="features", gpuMachines=["127.0.0.1,1"], labelsColumnName="labels", localHdfsMount=None, parallelTrain=True, username="sshuser", weightPrecision="float", workingDir="tmp"):
        """
        Set the (keyword only) parameters

        Args:

            brainScript (str): String of BrainScript config
            dataFormat (str): Transfer format (default: text)
            dataTransfer (str): Transfer strategy (default: local)
            featureCount (int): Number of features for reduction (default: 1)
            featuresColumnName (str): Features column name (default: features)
            gpuMachines (list): GPU machines to train on (default: [Ljava.lang.String;@17c8b8f0)
            labelsColumnName (str): Label column name (default: labels)
            localHdfsMount (str): Local mount point for hdfs:///
            parallelTrain (bool): Train using an MPI ring (default: true)
            username (str): Username for the GPU VM (default: sshuser)
            weightPrecision (str): Weights (default: float)
            workingDir (str): Working directory for CNTK (default: tmp)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setBrainScript(self, value):
        """

        Args:

            brainScript (str): String of BrainScript config

        """
        self._set(brainScript=value)
        return self


    def getBrainScript(self):
        """

        Returns:

            str: String of BrainScript config
        """
        return self.getOrDefault(self.brainScript)


    def setDataFormat(self, value):
        """

        Args:

            dataFormat (str): Transfer format (default: text)

        """
        self._set(dataFormat=value)
        return self


    def getDataFormat(self):
        """

        Returns:

            str: Transfer format (default: text)
        """
        return self.getOrDefault(self.dataFormat)


    def setDataTransfer(self, value):
        """

        Args:

            dataTransfer (str): Transfer strategy (default: local)

        """
        self._set(dataTransfer=value)
        return self


    def getDataTransfer(self):
        """

        Returns:

            str: Transfer strategy (default: local)
        """
        return self.getOrDefault(self.dataTransfer)


    def setFeatureCount(self, value):
        """

        Args:

            featureCount (int): Number of features for reduction (default: 1)

        """
        self._set(featureCount=value)
        return self


    def getFeatureCount(self):
        """

        Returns:

            int: Number of features for reduction (default: 1)
        """
        return self.getOrDefault(self.featureCount)


    def setFeaturesColumnName(self, value):
        """

        Args:

            featuresColumnName (str): Features column name (default: features)

        """
        self._set(featuresColumnName=value)
        return self


    def getFeaturesColumnName(self):
        """

        Returns:

            str: Features column name (default: features)
        """
        return self.getOrDefault(self.featuresColumnName)


    def setGpuMachines(self, value):
        """

        Args:

            gpuMachines (list): GPU machines to train on (default: [Ljava.lang.String;@17c8b8f0)

        """
        self._set(gpuMachines=value)
        return self


    def getGpuMachines(self):
        """

        Returns:

            list: GPU machines to train on (default: [Ljava.lang.String;@17c8b8f0)
        """
        return self.getOrDefault(self.gpuMachines)


    def setLabelsColumnName(self, value):
        """

        Args:

            labelsColumnName (str): Label column name (default: labels)

        """
        self._set(labelsColumnName=value)
        return self


    def getLabelsColumnName(self):
        """

        Returns:

            str: Label column name (default: labels)
        """
        return self.getOrDefault(self.labelsColumnName)


    def setLocalHdfsMount(self, value):
        """

        Args:

            localHdfsMount (str): Local mount point for hdfs:///

        """
        self._set(localHdfsMount=value)
        return self


    def getLocalHdfsMount(self):
        """

        Returns:

            str: Local mount point for hdfs:///
        """
        return self.getOrDefault(self.localHdfsMount)


    def setParallelTrain(self, value):
        """

        Args:

            parallelTrain (bool): Train using an MPI ring (default: true)

        """
        self._set(parallelTrain=value)
        return self


    def getParallelTrain(self):
        """

        Returns:

            bool: Train using an MPI ring (default: true)
        """
        return self.getOrDefault(self.parallelTrain)


    def setUsername(self, value):
        """

        Args:

            username (str): Username for the GPU VM (default: sshuser)

        """
        self._set(username=value)
        return self


    def getUsername(self):
        """

        Returns:

            str: Username for the GPU VM (default: sshuser)
        """
        return self.getOrDefault(self.username)


    def setWeightPrecision(self, value):
        """

        Args:

            weightPrecision (str): Weights (default: float)

        """
        self._set(weightPrecision=value)
        return self


    def getWeightPrecision(self):
        """

        Returns:

            str: Weights (default: float)
        """
        return self.getOrDefault(self.weightPrecision)


    def setWorkingDir(self, value):
        """

        Args:

            workingDir (str): Working directory for CNTK (default: tmp)

        """
        self._set(workingDir=value)
        return self


    def getWorkingDir(self):
        """

        Returns:

            str: Working directory for CNTK (default: tmp)
        """
        return self.getOrDefault(self.workingDir)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.CNTKLearner"

    @staticmethod
    def _from_java(java_stage):
        module_name=_CNTKLearner.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".CNTKLearner"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return _CNTKModel(java_model)


class _CNTKModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`_CNTKLearner`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.CNTKModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=_CNTKModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".CNTKModel"
        return from_java(java_stage, module_name)

